[CI] AbstractShapeGeometryFieldMapperTests testCartesianBoundsBlockLoader failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #5764 / ubuntu-2004-aarch64_checkpart1_platform-support-arm](https://gradle-enterprise.elastic.co/s/ufkvwizlx2m4o)
- [elasticsearch-pull-request #48863 / part-1](https://gradle-enterprise.elastic.co/s/rwcj54x4ur5ga)
- [elasticsearch-periodic-platform-support #5696 / amazonlinux-2023_platform-support-aws](https://gradle-enterprise.elastic.co/s/i6r5b27rys43c)
- [elasticsearch-periodic-platform-support #5696 / sles-12_platform-support-unix](https://gradle-enterprise.elastic.co/s/pmjiktexegc4g)
- [elasticsearch-periodic-platform-support #5681 / debian-11_platform-support-unix](https://gradle-enterprise.elastic.co/s/65qqahoblhmvc)
- [elasticsearch-periodic #5687 / release-tests](https://gradle-enterprise.elastic.co/s/kiazayhltgedi)
- [elasticsearch-periodic-platform-support #5671 / almalinux-8-aarch64_checkpart1_platform-support-arm](https://gradle-enterprise.elastic.co/s/u4uyudyyquj44)
- [elasticsearch-periodic #5682 / single-processor-node-tests](https://gradle-enterprise.elastic.co/s/cnx7gdssib3xk)
- [elasticsearch-periodic #5682 / concurrent-search-tests](https://gradle-enterprise.elastic.co/s/3577ihi2ofk2o)

**Reproduction Line:**
```
./gradlew ":server:test" --tests "org.elasticsearch.index.mapper.AbstractShapeGeometryFieldMapperTests.testCartesianBoundsBlockLoader" -Dtests.seed=A1DD0918EA3BE2CE -Dtests.locale=es-PY -Dtests.timezone=America/Chihuahua -Druntime.java=23
```

**Applicable branches:**
8.x

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.index.mapper.AbstractShapeGeometryFieldMapperTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!(testCartesianBoundsBlockLoader),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: geometry 'BBOX (-4.134783509883994E35, 5.472508646870402E35, 4.164420388635705E35, -5.704230026761559E35)' wasn't extracted correctly
Expected: "    BBOX (-4.134783509883994E35, 5.472508646870402E35, 4.164420388635705E35, -5.704230026761559E35)"
     but: was <BBOX (-5.807778858322827E35, 3.4028234663852886E38, 3.4028234663852886E38, -3.4028234663852886E38)>
```

**Issue Reasons:**
- [8.x] 9 failures in test testCartesianBoundsBlockLoader (1.0% fail rate in 874 executions)
- [8.x] 4 failures in pipeline elasticsearch-periodic-platform-support (22.2% fail rate in 18 executions)
- [8.x] 2 failures in pipeline elasticsearch-periodic (11.8% fail rate in 17 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/docs/changelog/119265.yaml
+++ b/docs/changelog/119265.yaml
@@ -1,0 +1,6 @@
+pr: 119265
+summary: Fix `AbstractShapeGeometryFieldMapperTests`
+area: "ES|QL, Geo"
+type: bug
+issues:
+ - 119201

--- a/docs/changelog/119265.yaml
+++ b/docs/changelog/119265.yaml
@@ -1,6 +1,6 @@
 pr: 119265
 summary: Fix `AbstractShapeGeometryFieldMapperTests`
-area: "ES|QL, Geo"
+area: "ES|QL"
 type: bug
 issues:
  - 119201

--- a/server/src/main/java/org/elasticsearch/index/mapper/AbstractShapeGeometryFieldMapper.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/AbstractShapeGeometryFieldMapper.java
@@ -124,7 +124,10 @@ public abstract class AbstractShapeGeometryFieldMapper<T> extends AbstractGeomet
 
                     private void read(BinaryDocValues binaryDocValues, int doc, GeometryDocValueReader reader, BytesRefBuilder builder)
                         throws IOException {
-                        binaryDocValues.advanceExact(doc);
+                        if (binaryDocValues.advanceExact(doc) == false) {
+                            builder.appendNull();
+                            return;
+                        }
                         reader.reset(binaryDocValues.binaryValue());
                         var extent = reader.getExtent();
                         // This is rather silly: an extent is already encoded as ints, but we convert it to Rectangle to

--- a/server/src/test/java/org/elasticsearch/index/mapper/AbstractShapeGeometryFieldMapperTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/AbstractShapeGeometryFieldMapperTests.java
@@ -11,7 +11,7 @@ package org.elasticsearch.index.mapper;
 
 import org.apache.lucene.document.Document;
 import org.apache.lucene.index.DirectoryReader;
-import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.LeafReader;
 import org.apache.lucene.store.Directory;
 import org.apache.lucene.tests.index.RandomIndexWriter;
 import org.apache.lucene.util.BytesRef;
@@ -30,6 +30,7 @@ import org.elasticsearch.test.hamcrest.RectangleMatcher;
 import org.elasticsearch.test.hamcrest.WellKnownBinaryBytesRefMatcher;
 
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.Optional;
 import java.util.function.Function;
 import java.util.function.Supplier;
@@ -61,7 +62,7 @@ public class AbstractShapeGeometryFieldMapperTests extends ESTestCase {
         Function<String, ShapeIndexer> indexerFactory,
         Function<Geometry, Optional<Rectangle>> visitor
     ) throws IOException {
-        var geometries = IntStream.range(0, 20).mapToObj(i -> generator.get()).toList();
+        var geometries = IntStream.range(0, 50).mapToObj(i -> generator.get()).toList();
         var loader = new AbstractShapeGeometryFieldMapper.AbstractShapeGeometryFieldType.BoundsBlockLoader("field", encoder);
         try (Directory directory = newDirectory()) {
             try (var iw = new RandomIndexWriter(random(), directory)) {
@@ -73,23 +74,39 @@ public class AbstractShapeGeometryFieldMapperTests extends ESTestCase {
                     iw.addDocument(doc);
                 }
             }
-            var indices = IntStream.range(0, geometries.size() / 2).map(x -> x * 2).toArray();
+            // We specifically check just the even indices, to verify the loader can skip documents correctly.
+            var evenIndices = evenArray(geometries.size());
             try (DirectoryReader reader = DirectoryReader.open(directory)) {
-                LeafReaderContext ctx = reader.leaves().get(0);
-                TestBlock block = (TestBlock) loader.reader(ctx).read(TestBlock.factory(ctx.reader().numDocs()), TestBlock.docs(indices));
-                for (int i = 0; i < indices.length; i++) {
-                    var idx = indices[i];
+                var byteRefResults = new ArrayList<BytesRef>();
+                for (var leaf : reader.leaves()) {
+                    LeafReader leafReader = leaf.reader();
+                    int numDocs = leafReader.numDocs();
+                    try (
+                        TestBlock block = (TestBlock) loader.reader(leaf)
+                            .read(TestBlock.factory(leafReader.numDocs()), TestBlock.docs(evenArray(numDocs)))
+                    ) {
+                        for (int i = 0; i < block.size(); i++) {
+                            byteRefResults.add((BytesRef) block.get(i));
+                        }
+                    }
+                }
+                for (int i = 0; i < evenIndices.length; i++) {
+                    var idx = evenIndices[i];
                     var geometry = geometries.get(idx);
                     var geoString = geometry.toString();
                     var geometryString = geoString.length() > 200 ? geoString.substring(0, 200) + "..." : geoString;
                     Rectangle r = visitor.apply(geometry).get();
                     assertThat(
                         Strings.format("geometries[%d] ('%s') wasn't extracted correctly", idx, geometryString),
-                        (BytesRef) block.get(i),
+                        byteRefResults.get(i),
                         WellKnownBinaryBytesRefMatcher.encodes(RectangleMatcher.closeToFloat(r, 1e-3, encoder))
                     );
                 }
             }
         }
+    }
+
+    private static int[] evenArray(int maxIndex) {
+        return IntStream.range(0, maxIndex / 2).map(x -> x * 2).toArray();
     }
 }

